File: Codes/1928-minimum-cost-to-reach-destination-in-time.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-10-03 12:42:02
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-10-03 12:51:57
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int minCost(int maxTime, vector<vector<int>>& edges, vector<int>& passingFees) {
        // vector<vector<pair<int, int>>> graph(passingFees.size());
        // for (vector<int>& edge : edges) {
        //     graph[edge[0]].push_back({edge[1], edge[2]});
        //     graph[edge[1]].push_back({edge[0], edge[2]});
        // }
        vector<vector<int>> dp(maxTime + 1, vector<int>(passingFees.size(), 10000000));
        dp[0][0] = passingFees[0];
        for (int t = 1; t <= maxTime; t++) {
            for (vector<int>& edge : edges) {
                int i = edge[0], j = edge[1], thisTime = edge[2];
                if (thisTime <= t) {
                    dp[t][j] = min(dp[t][j], dp[t - thisTime][i] + passingFees[j]);
                    dp[t][i] = min(dp[t][i], dp[t - thisTime][j] + passingFees[i]);
                }
            }
        }
        int ans = 10000000;
        for (int t = 0; t <= maxTime; t++) {
            ans = min(ans, dp[t].back());
        }
        return ans == 10000000 ? -1 : ans;
    }
};

File: Codes/1928-minimum-cost-to-reach-destination-in-time.py
'''
Author: LetMeFly
Date: 2024-10-03 12:52:33
LastEditors: LetMeFly
LastEditTime: 2024-10-03 12:55:56
'''
from typing import List

class Solution:
    def minCost(self, maxTime: int, edges: List[List[int]], passingFees: List[int]) -> int:
        dp = [[10000000] * len(passingFees) for _ in range(maxTime + 1)]
        dp[0][0] = passingFees[0]
        for t in range(1, maxTime + 1):
            for i, j, thisTime in edges:
                if thisTime <= t:
                    dp[t][j] = min(dp[t][j], dp[t - thisTime][i] + passingFees[j])
                    dp[t][i] = min(dp[t][i], dp[t - thisTime][j] + passingFees[i])
        ans = min(d[-1] for d in dp)
        return -1 if ans == 10000000 else ans